feat: Dispatch method (#1630)
software-mansion/TypeGPU: apps/typegpu-docs/src/content/docs/fundamentals/utils.mdx

---
title: Utilities
description: A list of various utilities provided by TypeGPU.
---

## Dispatch

The `prepareDispatch` function simplifies running simple computations on the GPU.
Behind the scenes, `prepareDispatch` wraps the callback into a `TgpuFn`, then it creates a compute pipeline and returns a dispatch function that runs the pipeline.

> This might help omit the serialization overhead for filling out buffers with initial data.

Author: Nice, thanks! I should have run those through gpt or something.

```ts twoslash
import tgpu, { prepareDispatch } from 'typegpu';
import * as d from 'typegpu/data';

const root = await tgpu.init();

const Boid = d.struct({
  index: d.u32,
  pos: d.vec3f,
});

// buffer of 2048 Boids
const boidsMutable = root.createMutable(d.arrayOf(Boid, 2048));

const dispatch = prepareDispatch({
  root,
  size: [2048] as const,
  callback: (x) => {
    // TGSL function body to run 2048 times on the GPU
    'kernel';
    const boidData = Boid({ index: x, pos: d.vec3f() });
    boidsMutable.$[x] = boidData;
  },
});

> dispatch().then(() => console.log("Dispatch finished!"));

Author: The api is now exactly like proposed, modulo the function name (`prepareDispatch`)

```

:::note
Remember to mark the callback with `'kernel'` directive to let TypeGPU know that this function is TGSL.
:::

> The returned dispatch function can be called multiple times.
> The pipeline is reused, so there is no extra overhead of successive calls.

Reviewer: ```suggestion
The returned dispatch function can be called multiple times.
Since the pipeline is reused, there’s no additional overhead for subsequent calls.
```
